hepstats/hypotests/calculators/basecalculator.py

```diff
@@ -436,3 +303,176 @@ def q(self, nll1: np.array, nll2: np.array, poi1, poi2,
+        ToysCalculator constructor with the toys loaded from a yaml file.
+
+        Args:
+            filename (str)
+            input : loss or fit result
+            minimizer : minimizer to use to find the minimum of the loss function
+            ntoysnull (int, default=100): minimum number of toys to generate for the null hypothesis
+            ntoysalt (int, default=100): minimum number of toys to generate for the alternative hypothesis
+            sampler : function used to create sampler with models, number of events and
+                floating parameters in the sample Default is `hepstats.fitutils.sampling.base_sampler`.
+            sample : function used to get samples from the sampler.
+                Default is `hepstats.fitutils.sampling.base_sample`.
+
+        Returns
+            ToysCalculator
+        """
+
+        calculator = cls(loss, minimizer, ntoysnull, ntoysalt, sampler, sample)
+        toysresults = calculator.toysresults_from_yaml(filename)
+
+        for t in toysresults:
+            calculator.add_toyresult(t)
+
+        return calculator
+
+    @property
+    def ntoysnull(self):
+        """
+        Returns the number of toys loss for the null hypothesis.
+        """
+        return self._ntoysnull
+
+    @property
+    def ntoysalt(self):
+        """
+        Returns the number of toys loss for the alternative hypothesis.
+        """
+        return self._ntoysalt
+
+    def _get_toys(self, poigen, poieval=None, qtilde=False, hypothesis="null"):
+        """
+        Return the generated toys for a given POI.
+
+        Args:
+            poigen (POI): POI used to generate the toys
+            poieval (POIarray): POI values to evaluate the loss function
+            qtilde (bool, optional): if `True` use the $$\tilde{q}$$ test statistics else (default) use
+                the $$q$$ test statistic
+            hypothesis: `null` or `alternative`
+        """
+
+        assert hypothesis in ["null", "alternative"]
+
+        if hypothesis == "null":
+            ntoys = self.ntoysnull
+        else:
+            ntoys = self.ntoysalt
+
+        ret = {}
+
+        for p in poigen:
+            poieval_p = poieval
+
+            if poieval_p is None:
+                poieval_p = POIarray(poigen.parameter, [p.value])
+            else:
+                if p not in poieval_p:
+                    poieval_p = poieval_p.append(p.value)
+
+            if qtilde and 0. not in poieval_p.values:
+                poieval_p = poieval_p.append(0.0)
+
+            poieval_p = asarray(poieval_p)
+
+            ngenerated = self.ntoys(p, poieval_p)
+            if ngenerated < ntoys:
+                ntogen = ntoys - ngenerated
+            else:
+                ntogen = 0
+
+            if ntogen > 0:
+                print(f"Generating {hypothesis} hypothesis toys for {p}.")
+                print(p, poieval_p)
+
+                self.generate_and_fit_toys(ntoys=ntogen, poigen=p, poieval=poieval_p)
```

Comment:
There is no safety net in case something goes wrong in, say, the fitting part?

Reply:
I could design something to count the number of fits that are bad or went wrong, and print a warning message above a certain threshold.